odoo/tools/cloc.py

```diff
@@ -0,0 +1,214 @@
+        s = re.sub(comments_re, replacer, s)
+        s = re.sub(r"\s*\n\s*", r"\n", s).lstrip()
+        return s.count("\n"), total
+
+    #------------------------------------------------------
+    # Enumeration
+    #------------------------------------------------------
+    def book(self, module, item='', count=(0, 0)):
+        if count[0] == -1:
+            self.errors.setdefault(module, {})
+            self.errors[module][item] = count[1]
+        else:
+            self.modules.setdefault(module, {})
+            if item:
+                self.modules[module][item] = count
+            self.code[module] = self.code.get(module, 0) + count[0]
+            self.total[module] = self.total.get(module, 0) + count[1]
+            self.max_width = max(self.max_width, len(module), len(item) + 4)
+
+    def count_path(self, path, exclude=None):
+        path = path.rstrip('/')
+        exclude_list = []
+        for i in odoo.modules.module.MANIFEST_NAMES:
+            manifest_path = os.path.join(path, i)
+            try:
+                with open(manifest_path, 'rb') as manifest:
+                    exclude_list.extend(DEFAULT_EXCLUDE)
+                    d = ast.literal_eval(manifest.read().decode('latin1'))
+                    for j in ['cloc_exclude', 'demo', 'demo_xml']:
+                        exclude_list.extend(d.get(j, []))
+                    break
+            except Exception:
+                pass
+        if not exclude:
+            exclude = set()
+        for i in exclude_list:
+            exclude.update(str(p) for p in pathlib.Path(path).glob(i))
+
+        module_name = os.path.basename(path)
+        self.book(module_name)
+        for root, dirs, files in os.walk(path):
+            for file_name in files:
+                file_path = os.path.join(root, file_name)
+
+                if file_path in exclude:
+                    continue
+
+                ext = os.path.splitext(file_path)[1].lower()
+                if ext in ['.py', '.js', '.xml']:
+                    if os.path.getsize(file_path) > MAX_FILE_SIZE:
+                        self.book(module_name, file_path, (-1, "Max file size exceeded"))
+                        continue
+
+                    with open(file_path, 'rb') as f:
+                        content = f.read().decode('latin1')
+                    if ext == '.py':
+                        self.book(module_name, file_path, self.parse_py(content))
+                    elif ext == '.js':
+                        self.book(module_name, file_path, self.parse_js(content))
+                    elif ext == '.xml':
+                        self.book(module_name, file_path, self.parse_xml(content))
+
+    def count_modules(self, env):
+        # Exclude standard addons paths
+        exclude_heuristic = [odoo.modules.get_module_path(m, display_warning=False) for m in STANDARD_MODULES]
```

Comment:
We symlink all enabled addons into the same folder, so an external tool takes care of which ones to symlink and which ones to not. It's easier than managing whole addon repos.

So, unless symlinks are expanded here, this will produce a false negative.

I know this might not be the way you guys deploy, but it has been perfectly valid until today, so I guess it wouldn't be so hard to add a flag like `--follow-symlinks` that takes into account the real path of an addon instead of the linked path. WDYT?

Reply:
If we just map the final module paths through `os.path.realpath` before processing them we should handle per-module symlinks transparently.

Of course that still doesn't support all deployment layouts e.g. if you physically move all modules into a single parent directory, including custom ones. But @antonylesuisse did not want to support those layouts at all, at least not in a first iteration. It's always possible to call the command manually for odd cases anyway.

Note: that exclusion heuristic was chosen for simplicity vs explicitly detecting official addons through the manifest or otherwise.